tests/functional/flows/test_create_flow.py

```diff
@@ -0,0 +1,220 @@
+    """
+    Return a list of identities for the provided principal set, updating the provided
+    existing_identities dict with any new identities that are created.
+
+    principal_set is a list of principals in the request
+    existing_identities is a dict mapping principal URNs to identities
+    """
+    identity_set = []
+    # Randomize the names for each principal set
+    available_identities = list(FLOW_IDENTITIES.keys())
+    shuffle(available_identities)
+
+    # Iterate over the principals in the request
+    for index, principal in enumerate(principal_set):
+        if principal in existing_identities:
+            # Use the existing identity if it's already been assigned
+            identity_set.append(existing_identities[principal])
+            continue
+
+        if principal not in SPECIAL_PRINCIPALS:
+            # Attempt to assign distinct identities to each principal
+            identity = get_identity(
+                principal.split(":")[-1],
+                available_identities[index % len(available_identities)],
+            )
+            identity_set.append(identity)
+            existing_identities[principal] = identity
+
+    return identity_set
+
+
+def value_for_field_from_output(name, output):
+    """
+    Return the value for a specified field from the output of a command.
+    """
+    match = re.search(rf"^{name}:[^\S\n\r]+(?P<value>.*)$", output, flags=re.M)
+    assert match is not None
+    return match.group("value")
+
+
+def test_create_flow_text_output(run_line):
+    # Load the response mock and extract metadata
+    response = load_response("flows.create_flow")
+    definition = response.metadata["params"]["definition"]
+    input_schema = response.metadata["params"]["input_schema"]
+    keywords = response.metadata["params"]["keywords"]
+    title = response.metadata["params"]["title"]
+    subtitle = response.metadata["params"]["subtitle"]
+    description = response.metadata["params"]["description"]
+    flow_administrators = response.metadata["params"]["flow_administrators"]
+    flow_starters = response.metadata["params"]["flow_starters"]
+    flow_viewers = response.metadata["params"]["flow_viewers"]
+
+    # Configure the owner identity
+    owner_identity = get_identity(response.json["flow_owner"].split(":")[-1], "pete")
+    identities = {response.json["flow_owner"]: owner_identity}
+    print(identities)
+
+    # Configure the identities for other roles
+    flow_administrator_identities = assign_identities_for_principal_set(
+        principal_set=flow_administrators,
+        existing_identities=identities,
+    )
+    flow_starter_identities = assign_identities_for_principal_set(
+        principal_set=flow_starters,
+        existing_identities=identities,
+    )
+    flow_viewer_identities = assign_identities_for_principal_set(
+        principal_set=flow_viewers,
+        existing_identities=identities,
+    )
```

Comment:
I found this stack of calls hard to follow at first because it's doing two data manipulations at once -- returning a list and updating the `identities` dict.

I think it might be easier to follow with a very small object to wrap it, like...

```python
class IdentityPool:
    IDENTITY_DATA = {"pete": ..., "monica": ..., ...}

    def __init__(self):
        self.identities = {}
        self.assigned_sets = {}

    def assign(self, set_name, principal_set):
        ...  # see usage below to guess at what this should do

...

pool = IdentityPool()
pool.assign("owner", [response.json["flow_owner"]])
pool.assign("administrator", response.json["flow_administrators"])
...
owner_identity = pool.assigned_sets["owner"][0]
flow_administrator_identities = pool.assigned_sets["administrator"]
...

load_response(
    RegisteredResponse(
        service="auth",
        path="/v2/api/identities",
        json={
            "identities": list(pool.identities.values()),
        },
    )
)
```

This is maybe a bit more prescriptive than I mean to be, but I'm just looking for ways to avoid having to juggle quite as many details as a reader. Splitting up the assignment from reading back the assigned values is the first thing which occurs to me as a way of doing that.

Reply:
Yeah, this seems reasonable to me. I'll take a quick swing at it.

Reply:
I've done something much along the lines of this suggestion that I think improves readability meaningfully. Could potentially improve further w/ more investment in the IdentityPool class, but have refrained from doing so unless/until we think it might prove useful for other tests. LMK what you think!